### ISSUE TITLE
Export go_toolchain_type in def.bzl

### ISSUE THREAD

**Issue by jayconrod** [bazelbuild/rules_go#630]
#620 changed how the toolchain rules return named providers. The symbol `go_toolchain_type` is now needed to access these, but it's defined in the private directory and not exported through def.bzl.

@ianthehat, it it safe to export this? Kubernetes used the toolchain in their `go_genrule`, but #620 broke them. I'd rather they not import `get_go_toolchain` from `go/private/common.bzl` directly, but that's essentially what they need.

cc @ixdy

**Comment by ianthehat** [comment]
It's not a public interface, and it will break them again when we switch to the new toolchain branch, but so long as they re okay with that, I don't mind them importing it directly.

**Comment by jayconrod** [comment]
Will the toolchain provider be stable enough in the future to have an exported interface, or will it always be internal?

**Comment by ixdy** [comment]
I'm not super excited about things breaking again (this is the second time changes in `rules_go` have broken our `go_genrule` rule), but working is better than perfect, I guess.

Our `go_genrule` is implemented in https://github.com/kubernetes/repo-infra/blob/master/defs/go.bzl and is used in a few places such as [here](https://github.com/kubernetes/kubernetes/blob/04c4c212266f3426fc770580b9d791fcef2531bb/staging/src/k8s.io/apimachinery/pkg/util/sets/BUILD#L25-L51).

Basically, we have some generator binaries which expect a Go tree to work properly, and so we have this elaborate dance to get things set up properly. It's possible we may be able to [refactor things so we don't need this](https://github.com/kubernetes/repo-infra/pull/19#issuecomment-313758322), but I don't have a full understanding of the space. (And I have only so much time for yak-shaving.)

If you have better suggestions on how to achieve what we're doing here I'd be happy to learn.

cc @mikedanese @spxtr

**Comment by ixdy** [comment]
FTR #623 broke us, not #620, though they are related PRs.

**Comment by ianthehat** [comment]
The go toolchain will be a stable interface, and usable by others.
You can see the changes coming in #625.

I have a prototype of something that attempts to build a valid temporary GOPATH full of links, so you that you can use it for running tools (notably `go vet` and `go fix` are high on my list of desirables).
`go vet` is easier, it only needs to read files, `go fix` has to update the original copies, which has more problems...

It works by propagating an aspect across the go_library tree to pick up all the source files consumed (including generated ones), and what their import paths should have been. 
This is intended as a temporary measure until the longer task of making the go/ packages understand varying file system layouts is ready.

I would be happy to work with you to finish that stuff off and then use it to re-implement your go_genrule, that way at least you it would be based on something we promise to keep working, longer term you will just be able to use the tool directly without a GOPATH and GOROOT.

**Comment by ixdy** [comment]
I'd like to bump our `rules_go` dependency but this issue is blocking that. Is there a workaround we can use before the full toolchain changes land?

**Comment by ianthehat** [comment]
Sorry for the delayed reply, I have been (and still am) on holiday.
For now just do what jay suggested and import the private symbol directly
```
load("@io_bazel_rules_go//go/private:go_toolchain.bzl", "get_go_toolchain")
```
and then use `get_go_toolchain(ctx)` in your rule.
That will keep working (for a while) with the new toolchains as well, as I swap the implementation of that function to the new form.

I will make sure you have a better (and more permanent) alternative before I remove that method, and will probably send you a PR to switch kubernetes to it.

**Comment by duginivijay** [comment]
Dear @ianthehat @jayconrod , I'm new to Bazel and I was trying to build an executable from MAC for Linux i.e's cross compilation.
I've read many threads and sub threads you guys created to get this issue fixed with toolchains, but I got so confused by reading multiple chains and now I'm unable to figure it out how could I solve the issue and implement toolchains rule to build y executable for Linux. Kindly could you please advice how can I solve my problem.

Threads I followed:
Issue 70, 79, 98, 142, 145, 625, 630, 663, pubref/rules_go, Platforms for Bazel document, Adds go_root rule.

**Comment by ianthehat** [comment]
You need pull request #683 (which is not submitted yet) and Bazel 0.5.4 (which is not released yet) and then compiling with --cpu=k8 on mac will build for linux
We are just waiting for the Bazel release of 0.5.4, which we are hoping will be very soon now, in order to submit the cross compilation work.
There is still more to do, but mac->linux is the one cross-compilation that is full supported to start with.

**Comment by duginivijay** [comment]
Thank you @ianthehat for your quick response. Kindly could you guide me how can we take this ahead ? like I read the thread there are 15 files which have been updated and the comments says we could use the bazel build command to generate the executable for linux. But to do that which all files should I update and how could I use or implement toolchain in to my project. Kindly could you please share your expertise knowledge.

**Comment by ianthehat** [comment]
The released Bazel 0.5.4 has toolchain support in it, and the tip of rules_go now has it all integrated, so now the only step needed to cross compile from mac to linux is to add `--cpu=k8` to your command line.

I am going to close this now, as the issue it originally referred to is no longer relevant, feel free to open a new issue if there was anything in here that you think was unresolved.
